### PR TITLE
Harmony 1069 - Provide friendly error message when query params are named incorrectly

### PR DESCRIPTION
**Pull request by vinnyinverso**

### CODE REVIEW

File: app/middleware/parameter-validation.ts [hunk 1]
```diff
@@ -27,6 +30,22 @@ function validateLinkTypeParameter(req: HarmonyRequest): void {
...
+ * @param req - The client request
+ */
+function validateCoverageRangesetParameterNames(req: HarmonyRequest): void {
+  const requestedParams = Object.keys(req.query);
+  const allowedParams = req.method == 'GET' ? coverageRangesetGetParams : coverageRangesetPostParams;
+  const incorrectParams = requestedParams.filter(param => !allowedParams.includes(param));
+  if (incorrectParams.length) {
+    const incorrectListString = listToText(incorrectParams, Conjunction.AND);
+    const allowedListString = listToText(allowedParams, Conjunction.AND);
+    throw new RequestValidationError(`Invalid query parameter(s) '${incorrectListString}'. Allowed parameters are ${allowedListString}`);
```
Comment (chris-durbin) [inline app/middleware/parameter-validation.ts:45]: Nice - I like the allowed parameters list. People might learn something!

File: app/middleware/parameter-validation.ts [hunk 2]
```diff
@@ -40,6 +59,9 @@ export default function parameterValidation(
 ): void {
   try {
     validateLinkTypeParameter(req);
+    if (req.url.match(RANGESET_ROUTE_REGEX)) {
```
Comment (chris-durbin) [inline app/middleware/parameter-validation.ts:62]: I tested the WMS route and see it allows bogus parameters still. Should we continue to ignore extra parameters there?

http://localhost:3000/C1233800302-EEDTEST/wms?SERVICE=WMS&VERSION=1.3.0&REQUEST=GetMap&BBOX=-180,-90,154.4170924291685765,90&CRS=CRS%3A84&WIDTH=2000&HEIGHT=1078&LAYERS=C1233800302-EEDTEST&STYLES&FORMAT=image%2Fpng&DPI=72&MAP_RESOLUTION=72&FORMAT_OPTIONS=dpi%3A72&TRANSPARENT=TRUE&granuleId=G1233800343-EEDTEST&turbo=true&foo=bar

**Review by chris-durbin — COMMENTED** [review]
I like the changes, just want to make sure we aren't breaking compatibility with the OGC coverages API. Asked a question in Slack about what OGC says to do with unrecognized parameters.

File: app/middleware/parameter-validation.ts [hunk 3]
```diff
@@ -27,6 +30,22 @@ function validateLinkTypeParameter(req: HarmonyRequest): void {
...
 }
 
+/**
+ * Validate that the req query parameter names are correct according to Harmony implementation of OGC spec.
+ *
+ * @param req - The client request
+ */
+function validateCoverageRangesetParameterNames(req: HarmonyRequest): void {
+  const requestedParams = Object.keys(req.query);
+  const allowedParams = req.method == 'GET' ? coverageRangesetGetParams : coverageRangesetPostParams;
```
Comment (indiejames) [inline app/middleware/parameter-validation.ts:40]: will `method` always be uppercase? Should this be a case-insensitive string compare?

File: test/ogc-api-coverages/get-coverage-rangeset.ts [hunk 4]
```diff
@@ -228,6 +228,16 @@ describe('OGC API Coverages - getCoverageRangeset', function () {
     });
   });
 
+  describe('when provided an incorrectly named set of parameters', function () {
+    StubService.hook({ params: { redirect: 'http://example.com' } });
+    hookRangesetRequest(version, collection, variableName, { query: { granuleId, outputCrz: '', maxResultz: 100 } });
+    it('rejects the request with an informative error message', function () {
+      console.log(this.res.text);
```
Comment (indiejames) [inline test/ogc-api-coverages/get-coverage-rangeset.ts:235]: Remove this debugging output

File: test/ogc-api-coverages/get-coverage-rangeset.ts [hunk 5]
```diff
@@ -228,6 +228,16 @@ describe('OGC API Coverages - getCoverageRangeset', function () {
...
+      expect(this.res.status).to.equal(400);
+      expect(this.res.text).to.include("Error: Invalid query parameter(s) 'outputCrz and maxResultz'");
```
Comment (indiejames) [inline test/ogc-api-coverages/get-coverage-rangeset.ts:237]: Maybe add a check that the allowed list (without being specific about what the list is) is given?

**Review by indiejames — APPROVED** [review]
Looks good. Just some minor comments.

File: app/middleware/parameter-validation.ts (hunk 2 above)
Comment (bilts) [reply app/middleware/parameter-validation.ts:62]: I'd prefer we reject them, but I'd be okay with this being another ticket if it's more than 1-2 points of effort.

**Comment by bilts** [comment]
> I like the changes, just want to make sure we aren't breaking compatibility with the OGC coverages API. Asked a question in Slack about what OGC says to do with unrecognized parameters.

I replied in Slack. I vote still reject. If it ever becomes problematic, we can fix it then, but right now it seems the benefits greatly outweigh the harm

**Review by chris-durbin — APPROVED** [review]

**Pull request merged**